### PR TITLE
[HIL-SERL] Add Flow Q-learning (FQL) agent with action chunking

### PR DESCRIPTION
**Pull request by jpizarrom**
## What this does

This PR adds Flow Q-learning (FQL) agent with action chunking QC-FQL

- add `ACFQLPolicy` https://arxiv.org/abs/2507.07969 inspired in https://github.com/isabelliu0/fql-pytorch and https://github.com/Dasyud/FQL-PyTorch
- update `ReplayBuffer` to get n_steps returns, based on https://github.com/DLR-RM/stable-baselines3/pull/2144 -> **this was removed and re-implemented following the JAX code from the paper**
- implement Warm-start RL WSRL https://arxiv.org/abs/2412.07762 -> configurable with`policy.offline_steps` for pretrain, `policy.online_step_before_learning` for warm

## How it was tested

- trained a policy on PandaPickCubeGamepad-v0 sim https://github.com/huggingface/lerobot/pull/1818#issuecomment-3406812063
- Added some tests

###  train a policy on PandaPickCubeGamepad-v0 sim 

charts of our run https://github.com/huggingface/lerobot/pull/1818#issuecomment-3406812063

sample config https://gist.github.com/jpizarrom/23985d4726e5ed29959af8741b01493e

This command will launch the learner to do 4k offline steps, then will wait for receiving 4k transitions from actor, and will start doing online RL.
```
uv run python -m lerobot.rl.acfql.learner \
  --config_path=../lerobot-configs-hil-serl/fql-newpipe/train_gym_hil_env_fql_lilkm_pushfreq25x20251007.json \
  --policy.offline_steps=4000 --policy.online_step_before_learning=4000
```

This command launch the actor, it will wait for the model paramaters, and then will start getting transitions
```
uv run python -m lerobot.rl.acfql.actor \
  --config_path=../lerobot-configs-hil-serl/fql-newpipe/train_gym_hil_env_fql_lilkm_pushfreq25x20251007.json \
  --policy.offline_steps=4000
```

In case the actor do not start getting transitions, and the learner wait for the actor's transitions. -> resume the actor from the learner checkpoint with `offline_steps=0`. The actor will start getting transitions right after loading the model from the checkpoint. Once the learner receives enough transitions, it  will start online RL, and also will start sending model params to the actor.

```
uv run python -m lerobot.rl.acfql.actor \
  --config_path=./outputs/.../checkpoints/last/pretrained_model/train_config.json \
  --resume=true --policy.offline_steps=0
```

### run tests

```bash
python -m pytest -sv ./tests/processor/test_acfql_processor.py
python -m pytest -sv ./tests/processor/test_observation_processor.py::test_process_single_image_cuda
python -m pytest -sv ./tests/rl/acfql/test_replay_buffer_n_steps.py
```

## TODO

- [x] train on sim PandaPickCubeKeyboard-v0
- [x] train on real robot arm SO100
- [x] train on real robot arm Franka Emika (with custom changes, not added to this PR)
- [ ] train on LeKiwi
- [x] update to the latest main, new pipeline processors and get policy based on JAX code from @s1lent4gnt  https://github.com/s1lent4gnt/lerobot/tree/lilkm/finetune-vla-qc-fql
- [x] use make_pre_post_processors
- [x] nstep sampling on buffer
- [x] add delta_wx,delta_wy,delta_wz configurable
- [x] WSRL
- [x] check if deltas rotation shall be a separated PR or make it configurable
- [x] train on sim PandaPickCubeGamepad-v0 new processors new pipeline
- [x] train on real robot arm SO100 pos + rot with new processors pipeline
- [x] gym_manipulator with gripper continuous configurable -> use `teleop.action_features`
- [x] training loop/buffer compatible with other policies SAC/ACFQL/ConRFT ... -> for now make a new directory `src/lerobot/rl/acfql`
- [x] refactor and clean the PR
- [x] tests for ReplayBufferNSteps
- [x] add normalization for gripper action [0,2], make gripper neutral action and mix/max bounds configurable
- ~~add_ee_pose_to_observation~~ -> it can be done in a future effort
- ~~fix recording with lerobot.rl.acfql.gym_manipulator~~ -> it is broken for the mail gym_manipulator it can be done in a future effort
- [ ] handle truncation on critic loss, maybe  Mask out truncated transitions

### CODE REVIEW

**Comment by s1lent4gnt** [comment]
Nice work @jpizarrom!
I have implemented Cal-QL here : https://github.com/s1lent4gnt/lerobot/blob/lilkm/port-conrft/src/lerobot/policies/conrft/modeling_conrft.py#L298 with this PR : #1832

**Comment by jpizarrom** [comment]
- This branch was updated to the latest processor pipelines
-  got a refactored policy following the original JAX code from the paper https://arxiv.org/abs/2507.07969, done by @s1lent4gnt https://github.com/s1lent4gnt/lerobot/tree/lilkm/finetune-vla-qc-fql

**Comment by jpizarrom** [comment]
this are the charts of training PandaPickCubeGamepad-v0 with `ccd4509f17ea8c7ed2f0062b29c192d8e0831333`

Screenshot from 2025-10-15 16-40-58

**Comment by jpizarrom** [comment]
Hi @michel-aractingi , this PR is ready to be reviewed, happy to receive your feedback :)

**Comment by jpizarrom** [comment]
an open point is that currently lerobot.rl.acfql.gym_manipulator uses gripper action `[-1,+1]`, this is the range used to save the datasets.

there can be two options:
- add support to gym-hil to allow actions [-1,+1]
- or to support gym-hil without changes, gym_manipulator shall be adjusted to use gripper [0,2]
- maybe parameterize the GripperConfig bounds and neutral_action, so they are not hard-coded on the gym_manipulator

File: src/lerobot/policies/acfql/configuration_acfql.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+    # Aggregation method for Q-values, can be "mean" or "max"
+    q_agg: str = "mean"
+    # Weight for the alpha parameter in the ACFQL algorithm
+    alpha: float = 10.0
+    # Number of steps for the flow in the ACFQL algorithm
+    flow_steps: int = 10
+    # Whether to normalize the Q-loss
+    normalize_q_loss: bool = False
+    # Whether to use TD loss (should be True for normal training)
+    use_td_loss: bool = True
```
Comment (michel-aractingi) [inline src/lerobot/policies/acfql/configuration_acfql.py]: use_td_loss param not referenced anywhere else

File: src/lerobot/policies/acfql/configuration_acfql.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+    # Weight for the critic target update
+    critic_target_update_weight: float = 0.005
+    # Aggregation method for Q-values, can be "mean" or "max"
+    q_agg: str = "mean"
+    # Weight for the alpha parameter in the ACFQL algorithm
+    alpha: float = 10.0
+    # Number of steps for the flow in the ACFQL algorithm
+    flow_steps: int = 10
+    # Whether to normalize the Q-loss
+    normalize_q_loss: bool = False
```
Comment (michel-aractingi) [inline src/lerobot/policies/acfql/configuration_acfql.py:178]: `normalize_q_loss` not used in the modeling script

File: src/lerobot/policies/acfql/modeling_acfql.py [hunk 3]
```diff
@@ -0,0 +1,1057 @@
...
+            use_target=False,
+            observation_features=observation_features,
+        )
+
+        q_vals = q_preds.mean(dim=0)
+        q_loss = -q_vals.mean()
+
+        # TODO (jpizarrom): make this configurable
+        lam = 1.0 / q_preds.abs().mean().detach()
+        q_loss = lam * q_loss
```
Comment (michel-aractingi) [inline src/lerobot/policies/acfql/modeling_acfql.py]:
```suggestion
        if self.config.normalize_q_loss:
            lam = 1.0 / q_vals.abs().mean().detach()
            q_loss = lam * q_loss
```
1. We should use the normalize_q_loss here?
2. Looking at the jax implementation: 
```
        qs = self.network.select('critic')(batch['observations'], actions=actor_actions)
        q = jnp.mean(qs, axis=0)

        q_loss = -q.mean()
        if self.config['normalize_q_loss']:
            lam = jax.lax.stop_gradient(1 / jnp.abs(q).mean())
            q_loss = lam * q_loss
```
They use the mean of the predictions, i.e., the`q_vals` instead of `q_preds` to normalize the loss.

File: src/lerobot/policies/acfql/configuration_acfql.py [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+    # Whether to use TD loss (should be True for normal training)
+    use_td_loss: bool = True
+    # Update-to-data ratio for the UTD algorithm (If you want enable utd_ratio, you need to set it to >1)
+    utd_ratio: int = 2
+    # Hidden dimension size for the state encoder
+    state_encoder_hidden_dim: int = 256
```
Comment (michel-aractingi) [inline src/lerobot/policies/acfql/configuration_acfql.py]: `state_encoder_hidden_dims` Also not used anywhere

File: src/lerobot/policies/acfql/configuration_acfql.py [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+    # Name of the vision encoder model (Set to "helper2424/resnet10" for hil serl resnet10)
+    vision_encoder_name: str | None = None
+    # Whether to freeze the vision encoder during training
+    freeze_vision_encoder: bool = True
+    # Hidden dimension size for the image encoder
+    image_encoder_hidden_dim: int = 32
+    # Whether to use a shared encoder for actor and critic
+    shared_encoder: bool = True
+    # Number of discrete actions, eg for gripper actions
+    num_discrete_actions: int | None = None
```
Comment (michel-aractingi) [inline src/lerobot/policies/acfql/configuration_acfql.py]: We can remove `num_discrete_actions` as I don't think we will use the discrete critic for the fql implmentation

**Review by michel-aractingi — COMMENTED** [review]
Great work so far! Great job with the tests also. Left a couple of comments for the config and modeling scripts

**Comment by jpizarrom** [comment]
Hi @michel-aractingi

- The PR was updated following you comments
- now gripper neutral action and mix/max bounds are configurable, with the default values needed by gym-hil, to do so `gym_manipulator_acfql` env with custom acfql env dataclasses was added
- the config sample https://gist.github.com/jpizarrom/23985d4726e5ed29959af8741b01493e was updated with a dataset with range [0,2] , dataset from @s1lent4gnt 
- other fixes e.g this one will show interventions in wandb again [env step info should not set IS_INTERVENTION](https://github.com/huggingface/lerobot/pull/1818/commits/0bb6e325ab09dd46efdbff9bf88c7fac97b547cb)[0bb6e32](https://github.com/huggingface/lerobot/pull/1818/commits/0bb6e325ab09dd46efdbff9bf88c7fac97b547cb)


if it is fine, there are some pending topics that can be addressed in a future effort, as they are related to the SAC HIL-SERL too
- add_ee_pose_to_observation
- fix recording for gym-hil on gym_manipulator

let me know if this also can be done later, or shall added to this PR
- handle truncation, maybe Mask out truncated transitions like in https://github.com/jpizarrom/lerobot/blob/87da651e2659b6640784a3e18139063e1cbe5374/src/lerobot/policies/acfql/modeling_acfql.py#L399
